File: Go/Reverse-Vowels-Of-A-String.go
package main

func main() {
	reverseVowels("hello")
}

func reverseVowels(s string) string {
	left, right := 0, len(s)-1

	sample := []rune(s)

	for left < right {
		for left < len(s) && !isVowel(sample[left]) {
			left++
		}
		for right >= 0 && !isVowel(sample[right]) {
			right--
		}

		if left < right {
			sample[left], sample[right] = sample[right], sample[left]
			left++
			right--
		}
	}
	return string(sample)
}

func isVowel(s rune) bool {
	vowels := []rune{'a', 'e', 'i', 'o', 'u', 'A', 'E', 'I', 'O', 'U'}

	for _, vowel := range vowels {
		if vowel == s {
			return true
		}
	}

	return false
}

File: Javascript/Reverse-Vowels-Of-A-String.js
/**
 *
 * @param {string} s
 * @return {string}
 */
const reverseVowels = (s) => {
  const vowels = new Set(["a", "e", "i", "o", "u"]);

  let [left, right] = [0, s.length - 1];

  let temp = s.split("");

  while (left < right) {
    while (left < s.length && !vowels.has(temp[left].toLowerCase())) {
      left++;
    }

    while (right >= 0 && !vowels.has(temp[right].toLowerCase())) {
      right--;
    }

    if (left < right) {
      let t = temp[left];
      temp[left] = temp[right];
      temp[right] = t;
      left++;
      right--;
    }
  }
  return temp.join("");
};
